hdat/reports.py

```diff
@@ -0,0 +1,57 @@
+import csv
+import sys
+
+from collections import defaultdict
+from itertools import chain
+
+
+def expand_key_parts(result, nested_key):
+    if nested_key[-1] == "*":
+        nested_data = get_nested_item(result, nested_key[:-1])
+        if isinstance(nested_data, dict):
+            for key in nested_data:
+                yield nested_key[:-1] + [key]
+    else:
+        yield nested_key
+
+
+def get_nested_item(result, nested_key):
+    num_levels = len(nested_key)
+    try:
+        if num_levels == 0 or nested_key[0] not in result:
+            return
+        elif num_levels == 1:
+            return result[nested_key[0]]
+        else:
+            return get_nested_item(result[nested_key[0]], nested_key[1:])
+    except TypeError:
+        error_msg = 'Result value "{}" is not iterable and has no specifier "{}"\n'
+        sys.stderr.write(error_msg.format(result, nested_key[0]))
+        return
+
+
+def print_results(results, input_keys_str):
+    if not input_keys_str:
+        input_keys_str = 'case_id,result_id,ran_on,commit,metrics.*'
+    reader_keys = csv.reader([input_keys_str])
+    input_keys = [key.strip() for key in list(reader_keys)[0]]
+    key_parts = [key.split('.') for key in input_keys]
+
+    keys = set()

```

Comment:
I think this is looking great.  Nice work!

The only thing I think could still be improved is the naming.  Some of the names (that I used in the snippet I gave you) are kind of vague.

E.g., maybe `report_data` is better than `all_data`?  Or maybe `all_result_data`?  Also, perhaps `expanded_keys_gen` should be `expanded_key_gens` (since there are multiple generators)?